src/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -13,17 +13,25 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
         {
             if (source == null) throw Error.ArgumentNull("source");
             IPartition<TSource> partition = source as IPartition<TSource>;
-            if (partition != null) return partition.ElementAt(index);
-            IList<TSource> list = source as IList<TSource>;
-            if (list != null) return list[index];
-            if (index >= 0)
+            if (partition != null)
+            {
+                bool found;
+                TSource element = partition.TryGetElementAt(index, out found);
+                if (found) return element;
+            }
+            else
             {
-                using (IEnumerator<TSource> e = source.GetEnumerator())
+                IList<TSource> list = source as IList<TSource>;
+                if (list != null) return list[index];
+                if (index >= 0)
                 {
-                    while (e.MoveNext())
+                    using (IEnumerator<TSource> e = source.GetEnumerator())
                     {
-                        if (index == 0) return e.Current;
-                        index--;
+                        while (e.MoveNext())
+                        {
+                            if (index == 0) return e.Current;
```

Comment:
It certainly sounds like it would be worth experimenting with. I'm currently in the middle of an emergency laptop repair that involves rather more duct tape than I'd like, so I won't be in a position to investigate this lunchtime, but you've piqued my curiosity.